plugin.go

```diff
@@ -0,0 +1,190 @@
+}
+
+func (p *Plugin) Init(log Logger) error {
+	p.log = log.NamedLogger(PluginName)
+
+	p.writersPool = sync.Pool{
+		New: func() any {
+			wr := new(writer)
+			wr.code = http.StatusOK
+			wr.data = make([]byte, 0, 10)
+			wr.hdrToSend = make(map[string][]string, 2)
+			return wr
+		},
+	}
+
+	return nil
+}
+
+func (p *Plugin) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		rrWriter := p.getWriter()
+		defer func() {
+			p.putWriter(rrWriter)
+			_ = r.Body.Close()
+		}()
+
+		next.ServeHTTP(rrWriter, r)
+
+		// if there is no X-Sendremotefile header from the PHP worker, just return
+		if path := rrWriter.Header().Get(xSendRemoteHeader); path == "" { //nolint:nestif
+			for k := range rrWriter.hdrToSend {
+				for kk := range rrWriter.hdrToSend[k] {
+					// re-add all headers from the worker
+					w.Header().Add(k, rrWriter.hdrToSend[k][kk])
+				}
+			}
+
+			// write original
+			w.WriteHeader(rrWriter.code)
+			if len(rrWriter.data) > 0 {
+				// write a body if exists
+				_, err := w.Write(rrWriter.data)
+				if err != nil {
+					p.log.Error("failed to write data to the response", zap.Error(err))
+				}
+			}
+
+			return
+		}
+
+		// we already checked that that header exists
+		path := rrWriter.Header().Get(xSendRemoteHeader)
+		// delete the original X-Sendremotefile header
+		rrWriter.Header().Del(xSendRemoteHeader)
+
+		// re-add original headers
+		for k := range rrWriter.hdrToSend {
+			for kk := range rrWriter.hdrToSend[k] {
+				// re-add all headers from the worker
+				w.Header().Add(k, rrWriter.hdrToSend[k][kk])
+			}
+		}
+
+		if !strings.HasPrefix(path, "http") {
+			http.NotFound(w, r)
+			return
+		}
+
+		ctx, cancel := context.WithCancel(context.TODO())
+		timer := time.AfterFunc(5*time.Second, func() {
+			cancel()
+		})
+
+		req, err := http.NewRequest("GET", path, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		req = req.WithContext(ctx)
+
+		resp, err := http.DefaultClient.Do(req)
+
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		defer func() {
+			_ = resp.Body.Close()
+		}()
+
+		if resp.StatusCode != http.StatusOK {
+			http.Error(w, http.StatusText(resp.StatusCode), resp.StatusCode)
+			return
+		}
+
+		var buf []byte
+		// do not allocate large buffer for the small files
+		if size := resp.ContentLength; size > 0 && size < int64(bufSize) {
+			// allocate based on provided content length
+			buf = make([]byte, size)
+		} else {
+			// allocate default buffer
+			buf = make([]byte, bufSize)
+		}
+
+		w.Header().Set(ContentTypeKey, ContentTypeVal)
+
+		for {
+			// Give 2 seconds to read response chunk
+			timer.Reset(2 * time.Second)
+
+			nr, er := resp.Body.Read(buf)
+
+			if nr > 0 {
+				nw, ew := w.Write(buf[0:nr])
+
+				if nw > 0 {
+					if f, ok := w.(http.Flusher); ok {
+						f.Flush()
+					}
+				}
```

Comment:
If you want to use `Flusher` it would be better to wrap a `ResponseWrite` with `http.NewResponseController`.